* İlk olarak, Sıcaklık verilerini Kaggle'dan indirin.
* Hedef değişkeninizin görünür sıcaklık ve sıcaklık arasındaki fark olduğu doğrusal bir regresyon modeli oluşturun. Açıklayıcı değişkenler olarak nem ve rüzgar hızı kullanın. Şimdi, modelinizi OLS kullanarak tahmin edin. Tahmin edilen katsayılar istatistiksel olarak anlamlı mıdır? Tahmini katsayılar önceki beklentileriniz doğrultusunda mı? Tahmin edilen katsayıları yorumlayınız. Hedef ve açıklayıcı değişkenler arasındaki ilişkiler nelerdir?
* Ardından, yukarıdaki modele nem ve rüzgar hızı etkileşimini dahil edin ve OLS'yi kullanarak modeli tahmin edin. Katsayılar istatistiksel olarak anlamlı mıdır? Nem ve rüzgar hızı için tahmini katsayıların işaretleri değişti mi? Tahmin edilen katsayıları yorumlayınız.

In [1]:
import numpy as np
import pandas as pd
from sklearn import linear_model
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

In [2]:
weather = pd.read_csv('./datasets/weatherHistory.csv')
weather.head(10)

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   
5  2006-04-01 05:00:00.000 +0200  Partly Cloudy        rain         9.222222   
6  2006-04-01 06:00:00.000 +0200  Partly Cloudy        rain         7.733333   
7  2006-04-01 07:00:00.000 +0200  Partly Cloudy        rain         8.772222   
8  2006-04-01 08:00:00.000 +0200  Partly Cloudy        rain        10.822222   
9  2006-04-01 09:00:00.000 +0200  Partly Cloudy        rain        13.772222   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   
5                  7.111111      0.85            13.9587   
6                  5.522222      0.95            12.3648   
7                  6.527778      0.89            14.1519   
8                 10.822222      0.82            11.3183   
9                 13.772222      0.72            12.5258   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   
5                   258.0          14.9569         0.0               1016.66   
6                   259.0           9.9820         0.0               1016.72   
7                   260.0           9.9820         0.0               1016.84   
8                   259.0           9.9820         0.0               1017.37   
9                   279.0           9.9820         0.0               1017.22   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.  
5  Partly cloudy throughout the day.  
6  Partly cloudy throughout the day.  
7  Partly cloudy throughout the day.  
8  Partly cloudy throughout the day.  
9  Partly cloudy throughout the day.

In [3]:
weather['Temperature_Diff (C)'] = weather['Temperature (C)'] - weather['Apparent Temperature (C)']

In [4]:
weather.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  Temperature_Diff (C)  
0  Partly cloudy throughout the day.              2.083333  
1  Partly cloudy throughout the day.              2.127778  
2  Partly cloudy throughout the day.              0.000000  
3  Partly cloudy throughout the day.              2.344444  
4  Partly cloudy throughout the day.              1.777778

In [5]:
weather.isnull().sum()

Formatted Date                0
Summary                       0
Precip Type                 517
Temperature (C)               0
Apparent Temperature (C)      0
Humidity                      0
Wind Speed (km/h)             0
Wind Bearing (degrees)        0
Visibility (km)               0
Loud Cover                    0
Pressure (millibars)          0
Daily Summary                 0
Temperature_Diff (C)          0
dtype: int64

In [6]:
weather.describe().T

count         mean         std        min  \
Temperature (C)           96453.0    11.932678    9.551546 -21.822222   
Apparent Temperature (C)  96453.0    10.855029   10.696847 -27.716667   
Humidity                  96453.0     0.734899    0.195473   0.000000   
Wind Speed (km/h)         96453.0    10.810640    6.913571   0.000000   
Wind Bearing (degrees)    96453.0   187.509232  107.383428   0.000000   
Visibility (km)           96453.0    10.347325    4.192123   0.000000   
Loud Cover                96453.0     0.000000    0.000000   0.000000   
Pressure (millibars)      96453.0  1003.235956  116.969906   0.000000   
Temperature_Diff (C)      96453.0     1.077650    1.678694  -4.811111   

                                  25%        50%          75%          max  
Temperature (C)              4.688889    12.0000    18.838889    39.905556  
Apparent Temperature (C)     2.311111    12.0000    18.838889    39.344444  
Humidity                     0.600000     0.7800     0.890000     1.000000  
Wind Speed (km/h)            5.828200     9.9659    14.135800    63.852600  
Wind Bearing (degrees)     116.000000   180.0000   290.000000   359.000000  
Visibility (km)              8.339800    10.0464    14.812000    16.100000  
Loud Cover                   0.000000     0.0000     0.000000     0.000000  
Pressure (millibars)      1011.900000  1016.4500  1021.090000  1046.380000  
Temperature_Diff (C)         0.000000     0.0000     2.216667    10.183333

In [7]:
Y = weather['Temperature_Diff (C)']
X = weather[['Humidity','Wind Speed (km/h)']]

lrm = linear_model.LinearRegression()
lrm.fit(X, Y)

print('Değişkenler: \n', lrm.coef_)
print('Sabit değer (bias): \n', lrm.intercept_)

Değişkenler: 
 [3.02918594 0.11929075]
Sabit değer (bias): 
 -2.438105415187706


In [8]:
import statsmodels.api as sm
Y = weather['Temperature_Diff (C)']
X = weather[['Humidity','Wind Speed (km/h)']]

X = sm.add_constant(X)
results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                             OLS Regression Results                             
================================================================================
Dep. Variable:     Temperature_Diff (C)   R-squared:                       0.288
Model:                              OLS   Adj. R-squared:                  0.288
Method:                   Least Squares   F-statistic:                 1.949e+04
Date:                  Mon, 30 Dec 2019   Prob (F-statistic):               0.00
Time:                          17:05:37   Log-Likelihood:            -1.7046e+05
No. Observations:                 96453   AIC:                         3.409e+05
Df Residuals:                     96450   BIC:                         3.409e+05
Df Model:                             2                                         
Covariance Type:              nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Tahmin edilen katsayılar istatistiksel olarak anlamlı mıdır? Tahmini katsayılar önceki beklentileriniz doğrultusunda mı? Tahmin edilen katsayıları yorumlayınız. Hedef ve açıklayıcı değişkenler arasındaki ilişkiler nelerdir?**

Modeli incelediğimizde hem Humidity parametresinin hem de Wind Speed (km/h) parameteresinin p değerinin 0'a çok yakın değerler aldığı görülüyor. O'a yakın olması iki parametrenin de tahmin için kullanılabilir olduğunu gösteriyor. Nem (Humuditiy) değerindeki artış Rüzgar hızındaki birim artışa göre tahmin edilen değerde daha büyük bir etki yaptığı görülüyor. (3.0292	> 0.1193).

**Ardından, yukarıdaki modele nem ve rüzgar hızı etkileşimini dahil edin ve OLS'yi kullanarak modeli tahmin edin.**

In [9]:
weather['Humidity_WindSpeed'] = weather['Humidity'] * weather['Wind Speed (km/h)']

Y = weather['Temperature_Diff (C)']
X = weather[['Humidity','Wind Speed (km/h)','Humidity_WindSpeed']]

lrm = linear_model.LinearRegression()
lrm.fit(X, Y)

print('Değişkenler: \n', lrm.coef_)
print('Sabit değer (bias): \n', lrm.intercept_)

Değişkenler: 
 [-0.17751219 -0.09048213  0.29711946]
Sabit değer (bias): 
 -0.08393631009786029


In [10]:
import statsmodels.api as sm
Y = weather['Temperature_Diff (C)']
X = weather[['Humidity','Wind Speed (km/h)','Humidity_WindSpeed']]

X = sm.add_constant(X)
results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                             OLS Regression Results                             
================================================================================
Dep. Variable:     Temperature_Diff (C)   R-squared:                       0.341
Model:                              OLS   Adj. R-squared:                  0.341
Method:                   Least Squares   F-statistic:                 1.666e+04
Date:                  Mon, 30 Dec 2019   Prob (F-statistic):               0.00
Time:                          17:12:22   Log-Likelihood:            -1.6669e+05
No. Observations:                 96453   AIC:                         3.334e+05
Df Residuals:                     96449   BIC:                         3.334e+05
Df Model:                             3                                         
Covariance Type:              nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity              -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)     -0.0905      0.002    -36.797      0.000      -0.095      -0.086
Humidity_WindSpeed     0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Katsayılar istatistiksel olarak anlamlı mıdır? Nem ve rüzgar hızı için tahmini katsayıların işaretleri değişti mi? Tahmin edilen katsayıları yorumlayınız.**

Modele Nem ve Rüzgar Hızı etkileşimi dahil edildiktan sonra diğer parametrelerin etkisi azaldı. Ayrıca etki katsayıları da (+) dan (-) değere dönüştü. 